_posts/2019-04-04-Improving-Parameter-Observers-Shogun.md

```diff
@@ -0,0 +1,208 @@
+a small dataset composed by two dimensional points which can have two labels `-1` and `1`.
+```python
+f_feats_train = csv_file("classifier_binary_2d_linear_features_train.dat")
+f_feats_test = csv_file("classifier_binary_2d_linear_features_test.dat")
+f_labels_train = csv_file("classifier_binary_2d_linear_labels_train.dat")
+f_labels_test = csv_file("classifier_binary_2d_linear_labels_test.dat")
+
+features_train = features(f_feats_train)
+features_test = features(f_feats_test)
+labels_train = labels(f_labels_train)
+labels_test = labels(f_labels_test)
+```
+The data are linearly separable, as we can see from the plot below.
+
+<p align="center">
+  <img src="/assets/img/binary_data_sample.png" />
+</p>
+
+We then initialize the model and we attach to it a parameter observer. In this example,
+we just used a logger which will print on stdout all the information it receives
+from the `perceptron` machine.
+```python
+perceptron = machine("AveragedPerceptron", labels=labels_train, learn_rate=1.0, max_iterations=1000)
+
+observer = parameter_observer("ParameterObserverLogger")
+perceptron.subscribe(observer)
+```
+
+We then train our model and we apply it to the test features.
+```python
+perceptron.train(features_train)
+labels_predict = perceptron.apply(features_test)
+```
+
+We are pretty much done! Once we run this code, the output will be something
+similar to this:
+
+![averaged perceptron example](/assets/img/averaged_perceptron_example.png)
+
+As you can see, the `AveragedPerceptron` emits its weights and bias.
+By looking at them we can get a certain idea about how the hyperplane is changing
+(if we are in a low dimensional space) and we can use those data to build nice
+visualizations. For instance, by adding more python code we can generate a nice
+video in which we show how the separation planes changes during training.
+
+```python
+#
+# Code for producing the animation
+#
+from matplotlib import pyplot as plt
+from matplotlib import animation
+import numpy as np
+
+fig = plt.figure()
+ax = plt.axes()
+line, = ax.plot([], [])
+
+plt.title("Binary Data Sample")
+
+# Plot the background for each frame
+def init():
+
+    plt.plot(features_train["x"], features_train["y"], 'bo', label="1")
+    line.set_data([], [])
+    line.set_label("hyperplane")
+
+    plt.xlabel("x")
+    plt.ylabel("y")
+    plt.legend(loc="upper left")
+
+    return line,
+
+# Print the observation line on screen
+def animate(i):
+    w = observer.get_observation(i).get("w")
+    bias = observer.get_observation(i).get("bias")
+
+    x = np.linspace(-10, 10, 1000)
+    line.set_data(x, x*w+bias)
+
+    return line,
+
+anim = animation.FuncAnimation(fig, animate, init_func=init, frames=len(w), interval=10, blit=True)
+anim.save('hyperplane.mp4', fps=15, extra_args=['-vcodec', 'libx264'], bitrate=500)
+
+```
+We can see the result in this image gif below. As you can see, the separation plane
```

Comment:
it is a bit weird that the thing converges within one iteration don't you think?
Maybe we could make the dataset a bit more complicated, so that finding a hyperplane requires some more iterations?

Reply:
Yeah, I was thinking the same actually. It would be better to have a more complicated dataset such to show also a better animation of it.

Reply:
It is pretty easy. Make a dataset that is linearly seperable, but just about so (thin margin). That should take longer. Or, just make one that isnt separable, then the algorithm wont converge